allennlp_models/generation/models/copynet_seq2seq.py

```diff
@@ -443,8 +450,24 @@ def _get_ll_contrib(
         gen_mask = (target_tokens != self._oov_index) | (target_to_source.sum(-1) == 0)
         log_gen_mask = (gen_mask + util.tiny_value_of_dtype(log_probs.dtype)).log().unsqueeze(-1)
         # Now we get the generation score for the gold target token.
-        # shape: (batch_size, 1)
-        generation_log_probs = log_probs.gather(1, target_tokens.unsqueeze(1)) + log_gen_mask
+        if self._label_smoothing is not None and self._label_smoothing > 0.0:
+            smoothing_value = self._label_smoothing / target_size
+            # Fill all the correct indices with 1 - smoothing value.
+            # shape: (batch_size, target_vocab_size + source_sequence_length)
+            one_hot_targets = torch.zeros_like(log_probs).scatter_(
+                -1, target_tokens.unsqueeze(1), 1.0 - self._label_smoothing
+            )
+            smoothed_targets = one_hot_targets + smoothing_value
```

Comment:
It seems like it would be faster to start with `torch.full_like(log_probs, smoothing_value)` and go from there? I did not measure it though.

`one_hot_targets` isn't used anywhere else, is it? It's not even properly "one hot" like this. Maybe it should be called `one_warm_targets` 🤣 .

Reply:
> It seems like it would be faster to start with `torch.full_like(log_probs, smoothing_value)` and go from there? I did not measure it though.

Something like 

```python
one_hot_targets = torch.full_like(log_probs, smoothing_value).scatter_(
    -1, target_tokens.unsqueeze(1), 1.0
)
```

I can do a little test to see if that's faster!

> `one_hot_targets` isn't used anywhere else, is it? It's not even properly "one hot" like this. Maybe it should be called `one_warm_targets` 🤣 .

Good point! I am stealing the variable `one_hot_targets` right from [here](https://github.com/allenai/allennlp/blob/cf113d705b9054d329c67cf9bb29cbc3f191015d/allennlp/nn/util.py#L821-L836). Another solution is just to use `smoothed_targets`:

```python
smoothed_targets = torch.zeros_like(log_probs).scatter_(
    -1, target_tokens.unsqueeze(1), 1.0 - self._label_smoothing
)
smoothed_targets = smoothed_targets + smoothing_value
```

or even 

```python
smoothed_targets = torch.zeros_like(log_probs).scatter_(
    -1, target_tokens.unsqueeze(1), 1.0 - self._label_smoothing
) + smoothing_value
```

but that might be putting too much logic into one line.

Reply:
Might be worth updating https://github.com/allenai/allennlp/blob/cf113d705b9054d329c67cf9bb29cbc3f191015d/allennlp/nn/util.py#L825-L828 to use this micro-optimization